### PR TITLE
Create script to convert all notebooks to MDX files

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/13875?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13875/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13875
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Write a script to convert all `.ipynb` files to `.mdx`. I took reference from [Langchain's implementation](https://github.com/langchain-ai/langchain/blob/a83357dc5ab5fcbed8c2dd7606e9ce763e48d194/docs/scripts/notebook_convert.py#L4).

There were some MDX syntax errors in the original notebooks (seems like sphinx can handle them/ignores them, but docusaurus can't). The edits to the notebook contents are to resolve these errors.

I've also renamed all notebooks to start with H1 so the titles will show up correctly. I could have made this part of the script but there's no good reason to not start the `.ipynb` files with an H1, so it seemed simpler to just do that.

The script uses nbconvert with a custom template so we can control how the code cells and output look. It's okay to leave them as default too, but by default the cell output has the same look as code cells, so it's a bit confusing.

Few things that can be done as follow up:
1. Convert the "download this notebook" links to an actual button
2. Edit the frontmatter of the pages to point to the ipynb files

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

preview docs job. feel free to browse through and demo some notebooks!

https://github.com/user-attachments/assets/d657714e-4c5a-4184-bcc1-154d6a365776

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb [hunk 1]
```diff
@@ -30,12 +16,13 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": null,
```
Comment (daniellok-db) [inline docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb:19]:
```suggestion
   "execution_count": 2,
```

File: docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb [hunk 2]
```diff
@@ -139,7 +126,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 6,
+   "execution_count": null,
```
Comment (daniellok-db) [inline docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb:129]:
```suggestion
   "execution_count": 6,
```

File: docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb [hunk 3]
```diff
@@ -665,7 +653,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 10,
+   "execution_count": null,
```
Comment (daniellok-db) [inline docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb:656]:
```suggestion
   "execution_count": 10,
```

File: docs/docs/llms/transformers/tutorials/fine-tuning/transformers-peft.ipynb [hunk 4]
```diff
@@ -9,9 +9,12 @@
   },
   {
    "cell_type": "raw",
-   "metadata": {},
+   "metadata": {
+    "vscode": {
+     "languageId": "raw"
+    }
+   },
```
Comment (daniellok-db) [inline docs/docs/llms/transformers/tutorials/fine-tuning/transformers-peft.ipynb:16]:
```suggestion
   "metadata": {},
```

File: docs/docs/llms/transformers/tutorials/fine-tuning/transformers-peft.ipynb [hunk 5]
```diff
@@ -1239,6 +1242,11 @@
     "display_table({\"prompt\": test_prompt, \"generated_query\": generated_query})"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": []
+  },
```
Comment (daniellok-db) [inline docs/docs/llms/transformers/tutorials/fine-tuning/transformers-peft.ipynb:1249]:
```suggestion
```

File: docs/docs/traditional-ml/hyperparameter-tuning-with-child-runs/notebooks/logging-plots-in-mlflow.ipynb [hunk 6]
```diff
@@ -85,11 +56,32 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": null,
```
Comment (daniellok-db) [inline docs/docs/traditional-ml/hyperparameter-tuning-with-child-runs/notebooks/logging-plots-in-mlflow.ipynb:59]:
```suggestion
   "execution_count": 2,
```

File: docs/docs/traditional-ml/hyperparameter-tuning-with-child-runs/notebooks/parent-child-runs.ipynb [hunk 7]
```diff
@@ -50,11 +34,19 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": null,
```
Comment (daniellok-db) [inline docs/docs/traditional-ml/hyperparameter-tuning-with-child-runs/notebooks/parent-child-runs.ipynb:37]:
```suggestion
   "execution_count": 2,
```

File: docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb [hunk 8]
```diff
@@ -139,7 +127,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 6,
+   "execution_count": null,
```
Comment (daniellok-db) [inline docs/docs/getting-started/logging-first-model/notebooks/logging-first-model.ipynb:130]:
```suggestion
   "execution_count": 6,
```

File: docs/scripts/nbconvert_templates/mdx/index.md.j2
Comment (daniellok-db) [inline docs/scripts/nbconvert_templates/mdx/index.md.j2:1]: this is the file that controls how different blocks are rendered

File: docs/scripts/convert-notebooks.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        body, _ = exporter.from_notebook_node(nb)
+        body = add_custom_component_imports(body)
+
+        with open(mdx_path, "w") as f:
+            f.write(body)
+
+        return mdx_path
+
+
+def main():
```
Comment (daniellok-db) [inline docs/scripts/convert-notebooks.py:72]:
```suggestion
def main(): 
```

File: docs/scripts/convert-notebooks.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
```
Comment (daniellok-db) [inline docs/scripts/convert-notebooks.py:72]:
```suggestion
def main():
```

File: docs/scripts/convert-notebooks.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    template_name="mdx",
+    extra_template_basedirs=["./scripts/nbconvert_templates"],
+)
+
+
+# add the imports for our custom cell output components
+def add_custom_component_imports(
+    body: str,
+) -> str:
+    return f"""import {{ NotebookCodeCell }} from "@site/src/components/NotebookCodeCell"
```
Comment (TomeHirata) [inline docs/scripts/convert-notebooks.py:51]: nit: is indent here correct?

**Comment by harupy** [comment]
Can we convert ipynb files to HTML files to render them in an iframe to isolate them?

File: docs/docs/llms/openai/notebooks/openai-embeddings-generation.ipynb [hunk 12]
```diff
@@ -133,7 +133,7 @@
     "   - Adds a space after specific HTML tags in a BeautifulSoup object for better text readability.\n",
     "\n",
     "2. **extract_text_from_url**: \n",
-    "   - Extracts text from a specified webpage section using its URL and a target ID. Filters and organizes the text from tags like <h>, <li>, and <p>, excluding certain irrelevant sections.\n",
```
Comment (daniellok-db) [inline docs/docs/llms/openai/notebooks/openai-embeddings-generation.ipynb:136]: markdown cells are copied as-is to docusaurus, so we can't include these as they are invalid (no closing tag). actually, even in non-MDX context, this is a problem:

Screenshot 2024-11-28 at 11 27 16 AM

File: docs/docs/llms/transformers/tutorials/fine-tuning/transformers-peft.ipynb [hunk 13]
```diff
@@ -1075,7 +1074,7 @@
    "source": [
     "### What's Logged to MLflow?\n",
     "\n",
-    "Let's briefly review what is logged/saved to MLflow as a result of your training. To access the MLflow UI, run `mlflow ui` commands and open https://localhost:PORT (PORT is 5000 by default). Select the experiment \"MLflow PEFT Tutorial\" (or the notebook name when running on Databricks) on the left side. Then click on the latest MLflow Run named `Mistral-7B-SQL-QLoRA-2024-...` to view the Run details.\n",
```
Comment (daniellok-db) [inline docs/docs/llms/transformers/tutorials/fine-tuning/transformers-peft.ipynb:1078]: seems like docusaurus attempts to parse this link as a real link, i get an error: `Cause: Can't parse URL https://localhost:PORT with base unspecified://`

**Review by serena-ruan — APPROVED** [review]
LGTM! One small question: what if there're some changes made to notebooks in the master branch? I hope git can figure out when merging

**Comment by daniellok-db** [comment]
@serena-ruan git seems to work well enough, but for edge cases i've just been handling these cases during weekly rebase (doesn't seem like it's a big issue)

**Pull request merged**